### PR TITLE
Feature: Totem Of Corruption Overlay & Effective Area

### PR DESCRIPTION
**Pull request by j10a1n15**
## What
This Pull Request adds a Totem Of Corruption Overlay, which includes the totem owner & the time left.
It also adds an option to hide the particles around the totem.
And it adds an effective area renderer, showing a sphere around the totem (Thank you Empa for figuring that out).
Also has an option to get reminded when the totem is about to run out (customiseable time)

<details>
<summary>Images</summary>

Overlay:
image
Effective Area:
image
Hidden Particles:
image

</details>

## Changelog New Features
+ Added Totem Overlay. - j10a1n15
    * Option to change the overlay distance.
    * Option to hide Totem Particles.
    * Option to show the effective area of a totem.
    * Option to get reminded when a totem is about to expire.

## Changelog Technical Details
+ Added drawSphereInWorld() method. - j10a1n15

### CODE REVIEW

**Review by hannibal002 — CHANGES_REQUESTED** [review]
see discord suggestions

File: src/main/java/at/hannibal2/skyhanni/features/fishing/TotemOfCorruption.kt [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+import java.awt.Color
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.seconds
+import kotlin.time.DurationUnit
+import kotlin.time.toDuration
+
+private val config get() = SkyHanniMod.feature.fishing.totemOfCorruption
+
+private var display = emptyList<String>()
+private var totems: List<Totem> = emptyList()
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/fishing/TotemOfCorruption.kt]: no reason for them to be outside of the class

File: src/main/java/at/hannibal2/skyhanni/features/fishing/TotemOfCorruption.kt [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+class TotemOfCorruption {
+
+    private val group = RepoPattern.group("features.fishing.totemofcorruption")
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/fishing/TotemOfCorruption.kt]: remove features

File: src/main/java/at/hannibal2/skyhanni/features/fishing/TotemOfCorruption.kt [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+                        && config.warnWhenAboutToExpire.toDuration(DurationUnit.SECONDS) > 0.seconds
+                    ) {
+                        playPlingSound()
+                        sendTitle("§c§lTotem of Corruption §eabout to expire!", 5.seconds)
+                    }
+                    Totem(totem.getLorenzVec(), timeRemaining, owner)
+                } else {
+                    null
+                }
+            }
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/fishing/TotemOfCorruption.kt]: move this part inside the getTotems method

File: src/main/java/at/hannibal2/skyhanni/features/fishing/TotemOfCorruption.kt [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    fun onTick(event: LorenzTickEvent) {
+        if (!event.repeatSeconds(2)) return
+        if (!isOverlayEnabled()) return
+
+        totems = getTotems()
+            .filterNotNull()
+            .mapNotNull { totem ->
+                val timeRemaining = getTimeRemaining(totem)
+                val owner = getOwner(totem)
+                if (timeRemaining != null && owner != null) {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/fishing/TotemOfCorruption.kt]: create an extra function for "is this a real totem" check. then you can use gettime remaining ?: return null and get owner ?: return null

File: src/main/java/at/hannibal2/skyhanni/features/fishing/TotemOfCorruption.kt [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+                    if (
+                        timeRemaining == config.warnWhenAboutToExpire.toDuration(DurationUnit.SECONDS)
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/fishing/TotemOfCorruption.kt]:
```suggestion
                        timeRemaining == config.warnWhenAboutToExpire.seconds
```

Comment (hannibal002) [reply src/main/java/at/hannibal2/skyhanni/features/fishing/TotemOfCorruption.kt]: Also, is time remaining always exactly one second? feels a bit weird to look at

File: src/main/java/at/hannibal2/skyhanni/config/features/fishing/TotemOfCorruptionConfig.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+import com.google.gson.annotations.Expose;
+import io.github.moulberry.moulconfig.annotations.ConfigEditorBoolean;
+import io.github.moulberry.moulconfig.annotations.ConfigEditorColour;
+import io.github.moulberry.moulconfig.annotations.ConfigEditorSlider;
+import io.github.moulberry.moulconfig.annotations.ConfigOption;
+
+public class TotemOfCorruptionConfig {
+
+    @Expose
+    @ConfigOption(name = "Show Overlay", desc = "Show the Totem of Corruption overlay." +
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/features/fishing/TotemOfCorruptionConfig.java:14]: This feature is the main toggle. either rename this to enabled, or notice in all other features below that this one needs to be enabled first.

File: src/main/java/at/hannibal2/skyhanni/features/fishing/TotemOfCorruption.kt [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+                }
+            }
+        }
+    }
+
+    @SubscribeEvent
+    fun onRenderWorld(event: LorenzRenderWorldEvent) {
+        if (!isEffectiveAreaEnabled()) return
+
+        val color = Color(SpecialColour.specialToChromaRGB(config.color), true)
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/fishing/TotemOfCorruption.kt]: use this utility function instead
```suggestion
        val color = config.color.toChromaColor()
```

**Review by hannibal002 — CHANGES_REQUESTED** [review]
not tested in code, looks already pretty good

File: src/main/java/at/hannibal2/skyhanni/features/fishing/TotemOfCorruption.kt (hunk 5 above)
Comment (j10a1n15) [reply src/main/java/at/hannibal2/skyhanni/features/fishing/TotemOfCorruption.kt]: > Also, is time remaining always exactly one second? feels a bit weird to look at

what do you mean

Comment (hannibal002) [reply src/main/java/at/hannibal2/skyhanni/features/fishing/TotemOfCorruption.kt]: is timeRemaining mod 1.seconds == 0.seconds?

**Review by hannibal002 — APPROVED** [review]
looks cool in game

**Pull request merged**